### DISCUSSION TITLE
Make Unity publishing easier

### DISCUSSION THREAD

**Discussion by reeseschultz** [genaray/Arch#143] [Ideas]
@genaray I've been integrating Arch into Unity (for reasons), and noticed something peculiar.

## Long-winded exposition
For Arch itself (**edit:** and non-codegen extensions), I only needed to pull these nuggets/transitive nuggets (2.7MB):
* Collections.Pooled
* CommunityToolkit.HighPerformance
* JobScheduler
* MessagePack
* Microsoft.Bcl.AsyncInterfaces
* Microsoft.Extensions.ObjectPool
* Microsoft.NET.StringTools
* System.Collections.Immutable
* System.Runtime.CompilerServices.Unsafe
* Utf8Json

And then Arch works beautifully. Note the absence of [Roslyn (Microsoft.CodeAnalysis)](https://github.com/dotnet/roslyn), despite the fact that `IIncrementalGenerator` is used in base Arch. So... how is codegen happening at all? Refer to https://github.com/sschmid/Entitas/issues/1005#issuecomment-1589865106. I'm assuming you must be intentionally using `4.1.0` for the same reason as the creator of Entitas (even though Unity's documentation is clearly wrong)? Or is this just a coincidence?

I figure the native Roslyn support was instigated by the extensive codegen in the Entities package for DOTS, [which I might be familiar with](https://github.com/reeseschultz/ReeseUnityDemos). The project I'm testing Arch in is not directly using DOTS (yet), but Roslyn is still generally accessible. We know that DOTS is invisibly subsuming old editor/runtime features to fend off tech debt, so this makes sense.

## The problem
All of that brings me to Arch.Extended. The codegen extensions are all doing this in their respective `csproj`:
```xml
<PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.1.0" PrivateAssets="all" />
```
Pay attention to `PrivateAssets="all"`. Base Arch codegen does *not* include that key-value pair. Keeping it in the extensions, or at all, is a problem for Unity users, assuming you want to support them (I think you should, because Arch can actually help them transition away from being directly dependent on its capabilities). And remember that Unity has never properly addressed dependency management, so that leaves three options for pulling Roslyn and its transitive dependencies:
* [UnityNuGet](https://github.com/xoofx/UnityNuGet) where Arch is not presently available in its artifact repository, so I did not test it
* [NuGetForUnity](https://github.com/GlitchEnzo/NuGetForUnity) which thrashes and crashes the editor repeatedly, and ends up not retrieving the dependencies it needs pertaining to SQLite (having to do with transitive `SQLitePCLRaw.batteries_v2` not being where no human nor machine would intuit)
* Manually, which is what I actually did because I hate myself

So what has to be pulled down when `PrivateAssets="all"`? About **31MB** of:
* Humanizer
* Microsoft.CodeAnalysis
* Microsoft.CodeAnalysis.CSharp
* Microsoft.CodeAnalysis.CSharp.Workspaces
* SQLitePCLRaw.batteries_v2 (which should actually come from [SQLitePCLRaw.bundle_green](https://www.nuget.org/packages/SQLitePCLRaw.bundle_green), whatever the hell that means
* SQLitePCLRaw.core
* SQLitePCLRaw.provider
* System.Composition.AttributeModel
* System.Composition.Hosting
* System.Composition.Runtime
* System.Composition.TypedParts
* System.IO.Pipelines
* System.Reflection.Metadata

Not only does this make direct Arch-Unity integration a massive pain, but also: what if I want to supply a UPM package that uses Arch.Extended as a dependency? (Which maybe I do.) Packing 2.7MB of DLLs into a package, with correct author/license attribution, is not awesome, but not horrible. The additional 31MB is pretty bad, though, especially when NuGet itself can't determine where one dependency is supposed to come from. Fortunately all of the nuggets are Apache 2.0 or MIT-licensed (mostly MIT), so redistribution is perfectly fine from a legal perspective either way.

## Concluding thoughts
Please consider my proposal to remove all instances of `PrivateAssets="all"` from your `csproj` files in the context of Roslyn. That's all I'm asking. Your users may end up using codegen anyway. Is it technically the right thing to expose your dependencies to them? Well, generally no, but this is a case where an exception should be made. It would seem I too am headed down codegen alley, despite regarding it as an anti-pattern in my past life as an enterprise dev. Seems like Rust is the only popular language properly geared for ECS due its trait system.

**Obligatory regards:** Arch is awesome. You're a veritable ECS prodigy. What you're doing is fantastic for Mono, Unity, Godot and other C# game devs, and you should be proud of yourself.

**Comment by genaray** [comment]
First of all, thank you for your interest in Arch, your kind words and this post here! :) 

The only question I have is, would removing PrivateAssets="all" have any disadvantage? Should it be removed completely or set to another value instead? And could it lead to potential problems?

**Reply by reeseschultz** [reply]
Good question. ~~According to [these docs](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets), there doesn't appear to be an alternative value for this situation.~~ *See below comment.*

The disadvantage in removing `PrivateAssets="all"` is that the exact Roslyn version may not be obvious to the consumer (of Arch), even though it is explicit in your `csproj`. Worst-case scenario: if a non-Unity consumer decides to do codegen with Roslyn without specifying its version, they may think they get Roslyn automagically (when it's in fact a version-specific transitive dep from Arch). Furthermore, if they attempt to target a feature for a mismatched Roslyn version, they may encounter compile-time or (more rarely) logic errors due to differing implementations.

That said, most people don't do codegen, probably wouldn't run into these issues regardless, and many .NET people use Visual Studio, which [illuminates transitive dependencies as an advertised feature](https://devblogs.microsoft.com/nuget/introducing-transitive-dependencies-in-visual-studio/). Most .NET people who use a library will call it out and its version in their `csproj`. When they do that, Arch's transitive dependency should not negatively impact their editor hints nor build process.

It is odd that Unity apparently passes Roslyn through to its users, or at least they do for the 2022.3 editor version. The exact version you're using in fact, I'm assuming because of the introduction of `IIncrementalGenerator`. But I don't know if Unity is intentionally handing down Roslyn to users or not, so I can't tell you for certain if they justified this scenario nor why. It kind of makes sense that they would for user-convenience, but Entitas does not depend on that. I did some digging and found that Entitas' separate codegen program, Jenny, includes the Roslyn DLLs via [releases](https://github.com/sschmid/Entitas/releases). Better safe than sorry, because what if the next Unity editor version upgrades Roslyn?

---

So after more consideration I'm leaning toward a different solution. Arch could specify `PrivateAssets="all"` with respect to Roslyn in all `csproj` files, but *convenient* Arch support for Unity would require offering up all required DLLs as a build artifact archive.

The process for achieving that cannot be opaque nor manual, mainly for security reasons. Automating the process, transparently verifying signatures from NuGet, etc., would be the right way to go about it IMO.

Unless you have a better idea, I could create a prototype for that in a separate repo as a GitHub Action. The goal: duplicating NuGet dependency resolution features specifically for Arch users who can't really use NuGet :laughing:. What do you think, @genaray?

---

**Edit:** I'm overcomplicating this. Actually, `dotnet publish` can handily bring in all required libraries. It'd be relatively trivial (compared to what I suggested above) to use it. Modifying the existing build script makes more sense to me.

And now I see this is related to #125...

**Reply by reeseschultz** [reply]
Sorry for blowing up your notifications, but good news: I've added (local) publishing for Unity to my Arch.Extended fork. Refer to: https://github.com/genaray/Arch.Extended/compare/master...reeseschultz:Arch.Extended:feature/unity-publish. Actually in this case we can do `PrivateAssets="analyzers"`.

I can do the same (locally) for Arch if you want. I can submit pull requests for both explaining everything, because even though there are very few changes, dealing with MSBuild is anything but straightforward. It turns out `dotnet publish` doesn't cut it. You need more control in order to force the target framework for all included projects. This also requires every `csproj` to have a `<TargetFramework>` attribute, even if it already has `<TargetFrameworks>` because... you know, MSBuild is terrible. Are you open to this approach?

After we work out local publishing, we can look at GitHub Actions. The only weirdness I foresee with that would be duplicative transitive DLLs between Arch and Arch.Extended which may confuse some users. Have to think about it.

**Edit:** I did [the same thing for Arch](https://github.com/genaray/Arch/compare/master...reeseschultz:Arch:feature/unity-publish), except I also added `PrivateAssets="analyzers"` to its codegen `csproj` (so pretty much the opposite of what I originally suggested in this discussion, lol).

**Reply by reeseschultz** [reply]
I figured it all out. The "answer" for this discussion is generating Unity-targeted assemblies via respective PRs for Arch (https://github.com/genaray/Arch/pull/150) and Arch.Extended (https://github.com/genaray/Arch.Extended/pull/24) without source generation (a SQLite dependency refuses to automatically resolve, and `System.Reflection.Emit` does not play nice with IL2CPP).

**Reply by genaray** [reply]
Thanks again! I have already merged the PRs. The only question I have is, does this not include the source-generated methods?

Arch.SourceGenerator in the main repo actually runs once when creating the e.g. nuget or just when building to generate some Arch methods. Among them e.g. all the generic overloads of methods. Does this no longer happen in the build script and thus these methods fall out? Or is the source-gen itself just not included in the output?

**Reply by genaray** [reply]
Unfortunately i was forced to remove `<TargetFramework>...</>` from the csproj since it forced Arch to use NetStandard 2.1. Atleast all the tests failed since the .Net6/7 API was not available anymore :/

**Reply by reeseschultz** [reply]
My bad.

I looked at it and it seems like we can get everything we want by setting `<TargetFramework>net7.0</TargetFramework>`.

All tests pass for that, and it appears `Directory.Build.targets` will still overwrite it with `netstandard2.1` but **only** for Unity publishing. I'll double-check that this is what's actually happening and submit a PR for both Arch and Arch.Extended as needed. Thanks for looking at this so quickly.

---

Also, I meant to say that the codegen project itself is not included in the output--I didn't explain that well. The generic overloads definitely still work--I'm using them in a Unity 2022.3 testbed.

However... the `EventBus` in Arch.Extended is a no-go for Unity publishing, because it needs on-the-fly codegen so far as I understand. It's not so much that the codegen can't happen at authoring time (in the editor), it's that the transitive dependencies of Roslyn include some SQLite libraries that just don't come through without manual intervention:
* [SQLitePCLRaw.core](https://www.nuget.org/packages/SQLitePCLRaw.core)
* [SQLitePCLRaw.provider](https://www.nuget.org/packages/SQLitePCLRaw.provider.e_sqlite3)
* [SQLitePCLRaw.batteries_v2](https://www.nuget.org/packages/SQLitePCLRaw.bundle_green)

So if we wanted `EventBus` to work I guess we could just like manually include those. I'd rather not do that, because it doesn't seem like the right way to go about it. Honestly, I don't know what's going on there (yet). What's more, though, is that if we do bring in all the transitive deps needed for Roslyn, then we have to cull a couple out, like `System.Threading` because 1) Unity already exposes that, and 2) it's not permissively licensed for redistribution. The redistribution thing only matters insofar as a future Unity publishing GitHub Action is concerned.

**Reply by reeseschultz** [reply]
For the issue with the `TargetFramework` I've submitted PRs https://github.com/genaray/Arch/pull/152 and https://github.com/genaray/Arch.Extended/pull/25.

**Reply by genaray** [reply]
Ah I realize, then I had just misunderstood ^^ The PRs look great, have just both merged, thanks! :) 

The EventBus is of course a pity, but it is also no must. In the end, it generates a construct through which you can send "events", ie data to receiving methods. To avoid boxing or lookups it uses roslyn, so it is as efficient as calling the receiving method directly and can even be inlined :)

**Reply by reeseschultz** [reply]
Thank you very much. I'm happy to help.

I read the wiki regarding the EventBus and I find it extremely impressive for the reasons you mentioned. Event-based systems have been of serious interest to me before I started doing game dev. So I'll try to get Roslyn working inside(!) Unity for a thing I'm doing, then ideally I'll figure out how to deal with the dependencies in a way reasonably distributable to users. If that same technique, assuming I find one, works well, then I'll report back with congruent changes for Arch.

**Reply by genaray** [reply]
I had a similar experience... I was looking for a clean and efficient way to trigger and receive events for a long time. The EventBus is probably as efficient as it gets. I'm curious if there is still a way found, keep me posted ^^

**Reply by reeseschultz** [reply]
Will do.

**Reply by genaray** [reply]
One more small question :) 
What is required to run the build script? It tells me that no project was found: `MSBUILD : error MSB1003: Specify a project or project map file. The current working directory does not contain a project or project map file.`

For it to work, the script probably needs to be copied into the correct dictionary, right? ^^ 
Edit: A small usage example would be great, so i can add it to the wiki :)

**Reply by reeseschultz** [reply]
You should be able to run `./scripts/UnityPublish.sh` from the terminal from the root directory of the project. Let me know if that works. It works for me.

**Reply by genaray** [reply]
Unfortunately it does not work on my end (windows). This error is thrown: 
```csharp
Projects to be restored are determined...
  All projects are up to date for recovery.
MSBuild version 17.4.0-preview-22470-08+6521b1591 for .NET
MSBUILD : error MSB1009: The project file does not exist.
Switch: t:Unity
```

Any idea why this happens? I was in the root and executed the .sh.

**Reply by reeseschultz** [reply]
Ah, probably because of this line in the script:

```sh
assemblyDir="`pwd`/dist/Assemblies"
```

Long ago, to be arbitrarily different, Microsoft decided Windows should use backslashes instead of forward slashes. So directory traversal on Windows is incompatible with pretty much every other operating system.

I recommend using [Git BASH](https://gitforwindows.org/), and then you should no longer have a problem with the slashes. Note that the older build script uses forward slashes just as I have with the Unity publishing script. Generally for automation we target Linux since it's cheap and secure when running scripts in the cloud, but you run into compatibility issues when running them locally.

**Reply by genaray** [reply]
Ah thank you! Funnily enough I ran the script with git bash but the error remains the same :D 
Pretty much out of ideas rn.
